csrc/device_lower/pass/grid_serialization.cpp

```diff
@@ -0,0 +1,177 @@
+//!
+//! We inspect the loop nest up to the point after which all outer loops are
+//! trivial. This corresponds to the outer-most loop in the generated
+//! kernel code.
+//!
+//! Conditions for using serial grid reduction:
+//!  - All reduction dimensions are either unparallelized or parallelized as
+//!    BID, not TID. Block and warp reductions could be allowed in the future,
+//!    but the current focus is on cases where all threads are doing separate
+//!    reductions simultaneously.
+//!  - rop is not an allreduce. Note that we could implement serial allreduce
+//!    but it would require inserting a separate grid sync after this outer
+//!    loop.
+//!  - There are no other reductions in this loop nest that are TID or BID
+//!    parallelized, unless they also satisfy the conditions above and their
+//!    reduction pattern matches this one. Otherwise our syncs will be
+//!    mismatched, and there is no good way to handle that yet.
+class GridSerializationSyncInserter : kir::ExprMutator {
+ public:
+  GridSerializationSyncInserter(const std::vector<Expr*>& exprs) {
+    kir::ExprMutator::traverseAndInsert(exprs);
+  }
+
+  static std::vector<Expr*> insert(const std::vector<Expr*>& exprs) {
+    GridSerializationSyncInserter inserter(exprs);
+    return inserter.exprs_;
+  }
+
+ private:
+  using kir::ExprMutator::dispatch;
+  using kir::ExprMutator::handle;
+
+  //! Record cur_expr_sync_pattern_ if this is a serial grid reduction
+  void handle(ReductionOp* rop) override {
+    if (rop->serialGridReductionRequested()) {
+      ParallelTypeBitmap sync_pattern;
+      auto out = rop->out()->as<TensorView>();
+      NVF_ERROR(out != nullptr);
+      for (int i : c10::irange((int)out->nDims())) {
+        IterDomain* ax = out->axis(i);
+        if (!ax->isReduction()) {
+          continue;
+        }
+        NVF_ERROR(
+            !ax->isThreadDim(),
+            "Serial grid reduction cannot be applied with block reductions: ",
+            rop->toString());
+        if (ax->isBlockDim()) {
+          sync_pattern.set(ax->getParallelType());
+        }
+      }
+
+      if (!sync_pattern.hasBID()) {
+        // Don't set cur_expr_sync_pattern_ since this is not actually a grid
+        // reduction
+        return;
+      }
+
+      if (cur_expr_sync_pattern_.has_value()) {
+        NVF_ERROR(
+            cur_expr_sync_pattern_.value() == sync_pattern,
+            "Reduction op ",
+            rop->toString(),
+            " has requested serial grid reduction, but pattern ",
+            sync_pattern.toString(),
+            " conflicts with previous pattern: ",
+            cur_expr_sync_pattern_.value().toString());
+      } else {
+        cur_expr_sync_pattern_ = sync_pattern;
+      }
+    }
+  }
+
+  void dispatch(Expr* expr) override {
+    if (auto loop = dynamic_cast<kir::ForLoop*>(expr);
+        cur_top_level_expr_ || (loop && loop->isTrivial())) {
+      // Never sync around trivial loops. Also avoid redefining
```

Comment:
Placing sync around a trivial loop results in an error?

Reply:
`cur_top_level_expr_` is the expression we'll place syncs around in `insertSyncs`. This check just bypasses setting `cur_top_level_expr_`, so if there are trivial loops, we will still dispatch into them, but they won't count as top-level expr.